### ISSUE TITLE
Go Implementation using Capnproto

### ISSUE THREAD

**Issue by lthibault** [ipld/ipld#87]
Hello,

As per the notice in the README, I am opening an issue to discuss my plans to work on an equivalent version of [go-ipld-cbor](https://github.com/ipfs/go-ipld-cbor) using [CapnProto](https://capnproto.org/).

Does anybody have a rough sketch of how to approach this?  A recommended roadmap, perhaps?

**Comment by rvagg** [comment]
/cc @warpfork 

@lthibault sorry for the lack of responsiveness here, I'm not sure why I didn't see this originally. Mostly we're active in the specs repo and this would be a great place to do that because a spec doc in https://github.com/ipld/specs/tree/master/block-layer/codecs would help it be more universally useful. I suspect we need to update the README here.

What you should do for Go is make something work for https://github.com/ipld/go-ipld-prime, that's where all of the current IPLD Go work is doing and all of the existing libraries are mostly in maintenance-mode. It's under quite a bit of flux as @warpfork does perf work and tries to land on the most useful interfaces given the tradeoffs being made.

https://github.com/ipld/go-ipld-prime-proto _might_ be a useful example of how to go about this. It implements DAG-PB for go-ipld-prime.

**Comment by rvagg** [comment]
I should also have mentioned that DAG-CBOR and DAG-JSON are both implemented inside of go-ipld-prime, so you can look at them as examples too. They're in the repo but need to be explicitly added in to any project that wants to make use of them, so they could just as easily be external projects like go-ipld-prime-proto.

**Comment by lthibault** [comment]
Hi @rvagg, thanks for the info - this is very helpful!

It sounds like there are two parts to this:

- [ ] Create a spec and submit a PR to https://github.com/ipld/specs
- [ ] Create an implementation that satisfies https://github.com/ipld/go-ipld-prime

If that's the case, 👍.

>I should also have mentioned that DAG-CBOR and DAG-JSON are both implemented inside of go-ipld-prime [...] They're in the repo but need to be explicitly added in to any project that wants to make use of them.

I'm not sure I follow. DAG-CBOR seems to be implemented as a [stand-alone repo](https://github.com/ipfs/go-ipld-cbor).  I don't see any implementation in go-ipld-prime.

There's [go-ipld-prime/codec/dagcbor](https://github.com/ipld/go-ipld-prime/tree/master/codec/dagcbor), but it doesn't appear to be a full implementation of DAG-CBOR.  (In fact, could you help clarify what relationship is between this package and go-ipld-cbor?)

**Comment by warpfork** [comment]
Hi hi!

So first of all about the repos we have: in short: `go-ipld-prime` is the new; `go-ipld-cbor` and `go-ipld-format` and friends are the old.

We started the `go-ipld-prime` repo fresh when we started to get increasingly clear in our specs about [the Data Model](https://github.com/ipld/specs/blob/master/data-model-layer/data-model.md).  The [Node](https://godoc.org/github.com/ipld/go-ipld-prime#Node) interface is directly correlated with the concepts of the Data Model.  We made a new repository for this so that we make conceptually clean code without fear; and so that other projects can migrate at leisure rather than be forced by upstream release treadmills, since we knew we'd be making large structural changes and those aren't always easy to adopt.  

(There were other reasons as well: We also had significant problems with consistency in the older repos (the cbor repo had json related methods for debugging?  would someone expect all codecs to implement these?  why?  many issues like this...), and significant issues with performance (I cannot advise that anyone look at how many allocations are created in some of the common operations in `go-ipld-cbor` without giving a "are you in a state to receive information that might hurt you"-style disclaimer) that could only be addressed by deeply significant changes to both internals as well as the public API.)

At this point, the `go-ipld-prime` repo has significantly more features than the older repos -- the [traversal system](https://godoc.org/github.com/ipld/go-ipld-prime/traversal) and [IPLD Selectors](https://github.com/ipld/specs/tree/master/selectors) for example exist in `go-ipld-prime` _only_ (and we will _not_ be backporting them to the older `go-ipld-format` and `go-ipld-cbor` repos, because doing so would require... well, turning them into `go-ipld-prime`). 

To sum it up: we strongly suggest any new code be developed looking at `go-ipld-prime`, and not the other repos.  That we haven't put warning signs in the readmes of the other repos is a failing (... mine, really.  I should get on that).

And yes, go-ipld-prime/codec/dagcbor is a full implementation of DAG-CBOR :)  The relationship of it and `go-ipld-cbor` is: the former completely replaces the latter.  If you're not finding things you expected to see, maybe they're just in one of the other related packages in go-ipld-prime?  Or, we have some helper methods that we need to add somewhere (I'm not sure there's a "WrapObject" equivalent yet, but we could probably add that pretty fast).

**Comment by lthibault** [comment]
Hi @warpfork ,

First of all, thank you for this very detailed, helpful response.  I'm continually impressed by the community's warm welcome and thoughtfulness.  It's truly a pleasure to be here.  (@rvagg this gratitude extends to you as well!)

Ok, so your general point is well taken:  I should look at go-ipld-prime and essentially discount the rest.  Out of sheer curiosity, why the name "prime"?  I found it a bit perplexing, but I'm known to read into things too much ...

**Comment by warpfork** [comment]
About a new IPLD codec in general, and about a captnproto codec in particular:

The key definition of an IPLD Codec is: it turns serial data into something we can recognize as Data Model; and it turns Data Model into something serial.

This image hasn't made it into our documentation yet (we're still fussing about the details) but it may be helpful:

![](https://user-images.githubusercontent.com/495647/76918925-897e2300-691b-11ea-9fda-b697bece97e3.png)

So what exactly this entails depends on the codec.

Some codecs, like DAG-CBOR, can encode **anything** from the Data Model: any map key in the Data Model can be a map key in DAG-CBOR: hooray.

Some codecs, like (to take a fairly extreme example) *git*, handle this by "pretending" a git commit object is a map, with certain well-known keys like "author_name", "author_date", "blob", etc.  This means we definitely can't take arbitrary IPLD Data Model content and serialize it with the git codec!  But, we can deserialize git content, get Data Model, use any of our other code on that (like traversals or IPLD Selectors!), and also (if we didn't add anything non-git-like to it), serialize it again without loss.  So there are major caveats around the git codec, but it does useful things.

Trying to build a codec for capnproto might raise some interesting questions.  If I understand correct, capnproto has a fairly high amount of internal structure (you can't put just anything in it: there are some expected structure to the content... and, the serialization format and the structural expectations are conmingled).  And unlike git -- which has internal structure, but there's really only *one* internal structure and it's quite well known, so we just call it "git" and that's the end of it -- capnproto can have *many* internal structures, and you need *a capnproto schema* in order to understand that internal structure (it's not a self-describing format).

Things with internal structure but non-self-describing formats are tricky.  We can write codec libraries for them: but, inevitably, that codec library ends up having a constructor function somewhere that *takes a schema document specific to that format*.  Which is perfectly fine _in code_... but, doesn't play very cleanly with the idea of multicodecs, where a single varint is all the entropy we have for describing what codec to use.  And we expect multicodec numbers to be able to fully describe what codec code/configuration is needed to process some data, because otherwise we can't make traversals that cross CID links work transparently.

I don't actually have guidance for you on what to do about this.

**Comment by warpfork** [comment]
(...sorry for the close, that was an accident; I have incredibly fat fingers today?)

**Comment by warpfork** [comment]
>  Out of sheer curiosity, why the name "prime"?

I'm perennially bad at naming things :sweat:   I think at the time I must have been visualizing math notation like `A` and `A'`.

We'd probably like to rename some of the libraries, someday, to try to reduce confusion going forward.  Golang package imports don't make that easy, though, so it's a discussion we've avoided so far :weary:

**Comment by lthibault** [comment]
@warpfork Thanks for the visual - the good news is that it matches the mental model I've built from reading various docs/specs!

>If I understand correct, capnproto has a fairly high amount of internal structure (you can't put just anything in it: there are some expected structure to the content... and, the serialization format and the structural expectations are conmingled).

Yes.  In this respect it's much like Protocol Buffers, but capnp has quite a few useful tricks up its sleeves for reconciling its schema-oriented design with dynamic data.  Without going into too much detail, capnp provides:

- `AnyPointer` (which is semantically equivalent to Go's `interface{}`)
- Generics (which is equivalent to Go's ... erm ... nevermind... 😉)

>You need a capnproto schema in order to understand that internal structure (it's not a self-describing format)

How does https://github.com/ipld/go-ipld-prime-proto solve this issue (at a high level)?

I wonder if we're not speaking from two different levels of analysis.  Perhaps my initial posts were unclear/confused, but I'm increasingly thinking from a Layer-0 perspective.  My current strategy is to implement a block in capnp, as such:

```capnp
Multihash struct {
    htype @0 :UInt64;
    value @1 :Data;
}

CID struct {
    codec @0 :UInt64;
    multihash @1 :Multihash;
}

Block struct {
    cid @0 :CID;
    data @1 :Data;
}
```

From there, it seems like I can implement a `Node` type that can handle arbitrary linked blocks (including those encoded with a completely different codec)... unless I'm overlooking something?

I think a full-fledged codec is still doable, but I'm kind of pushing this back for the time being.

> Things with internal structure but non-self-describing formats are tricky. We can write codec libraries for them: but, inevitably, that codec library ends up having a constructor function somewhere that takes a schema document specific to that format.

When you say "things", are you referring to high-level objects like Git, or low-level things like blocks?

>I think at the time I must have been visualizing math notation like A and A'.

Makes sense 👍

**Comment by rvagg** [comment]
> How does https://github.com/ipld/go-ipld-prime-proto solve this issue (at a high level)?

Yeah, so DAG-PB is definitely _not_ a good example of anything that we're trying to pursue here. We sort of consider it a previous generation thing that we're stuck with because it's baked into UnixFSv1 for IPFS. It's maintained because the data is heavily in use, but nothing new that we're doing is relying on it. DAG-PB is only "DAG" in that it has a means for links, but it doesn't support the full Data Model and instead has a very restricted schema of what it can include, laid out nicely now at https://github.com/ipld/specs/blob/master/block-layer/codecs/dag-pb.md. The question for you in your schema is what can `data` include? Is this just a byte array format for distributing raw bytes? Is there anything else you could do with this format in terms of the Data Model? If it's just a byte array then you've only got a single element from the Data Model: `bytes`. Without a `link` type it's not even a "DAG" format. Perhaps that's all you're aiming for, but you did raise both `capnproto` and `dag-capnproto` in the multicodecs PR so we'd need to figure out where the DAG is in the latter.

We've been discussing the need for more clarity about this stuff, I think @wapfork might even have the beginnings of a table somewhere that describe where all the vairous codecs sit in terms of their support of the Data Model. We have a wide range, from DAG-CBOR and DAG-JSON which support the entirety of the Data Model structured in fairly arbitrary ways, down to Git and Bitcoin which are formats that are fixed and have a very specific schema that they bring with them and we shoehorn them in to the Data Model. But we also have DAG-PB which supports a subset of the Data Model and includes native `links` and some amount of arbitrary structure (mainly in the linking), and we also have plain CBOR and JSON—the former supports the full Data Model except for links and the latter supports the Data Model except for links and bytes. So you'll want to consider how far along that line you can push. The more toward arbitrary structure the better of course but it would seem like you're limited in ways similar to DAG-PB (fwiw I'm not sure we'd name it "DAG" if it were new and the current IPLD team were responsible for naming it today).

**Comment by lthibault** [comment]
@rvagg I'm beginning to suspect there's some latent misunderstanding on my part.  Please bear with me! 🙂 

>The question for you in your schema is what can data include? 
>[...] dag-capnproto in the multicodecs PR [...]

To clarify:  my aim is definitely to support a full-fledged DAG.

In other words:

- I'm aiming for feature parity with the CBOR implementation in go-ipld-prime
- I'm primarily interested in implementing `Kind`s (map, list, etc.)

> Is this just a byte array format for distributing raw bytes? 

Capnp's `Data` type is indeed an array of raw bytes, but isn't that exactly what [the block spec](https://github.com/ipld/specs/blob/master/block-layer/block.md) says it should be?

I understand that `block != dag`, but my impression was that a dag was _composed_ of linked blocks (= `node`).

This brings up a good point, though.  In reading through go-ipld-prime, I see no mention of blocks at all, whereas my understanding was that nodes are composed of linked blocks.  I'm guessing the approach is to use native Go memory-structures to manipulate nodes in memory (implies:  no notion of blocks), and to marshal said structures into linked blocks at serialization-time.

Have I understood correctly?  If so, where does [go-block-format](github.com/ipfs/go-block-format) come in (or is it also deprecated)? 

Relatedly:

- where and how does go-ipld-prime interact with a datastore?  
- is [go-blockservice](github.com/ipfs/go-blockservice) still relevant?

The reason I ask is because Capnp doesn't distinguish between memory and wire representations of data; there is no serialization step.  Instead, the `Node` (or `Map`, or `Bool`) memory object can be passed around directly (e.g. over bitswap).  This is the behavior I'm ultimately interested in.

**Comment by mikeal** [comment]
> DAG-PB is definitely not a good example of anything that we're trying to pursue here

I’ll try to clarify this a bit more. `dag-pb` was developed before IPLD existed and before anyone had really figured out the best way to compose content addressed data structures.

IPLD “native” codecs support the full [IPLD Data Model](https://github.com/ipld/specs/blob/master/data-model-layer/data-model.md). The way `dag-pb` is constructed doesn’t quite do that. Supporting the full data model is highly recommended if you’re doing a new codec and it’s the primary difference between “old” and “new” codecs.

One remaining but important difference is how links are conceptualized. In the Data Model a link is just a `CID`, it’s actually much easier to implement and reason about. You can sort of ignore how links are built in `dag-pb` for anything new, just find some way to tag some binary as being a “link” and stick the serialized CID in there.

**Comment by mikeal** [comment]
> where and how does go-ipld-prime interact with a datastore?

In general, IPLD doesn’t have a strong opinion about how things are stored. In IPLD everything at some point turns into a “Block” that is referenced by `CID` so all you ever need is a key/value store for the storage component.

All the fancy things we do in storage and in the services built around storage are practically invisible to IPLD. You shouldn’t need to care about the storage or network layer at all when you write a codec.

**Comment by mikeal** [comment]
> Capnp's Data type is indeed an array of raw bytes, but isn't that exactly what the block spec says it should be?

We literally just brought up in our team call on Monday that the spec doc for Block is “confusing” so my apologies on this one ;)

This is how it breaks down.

IPLD references “a series of bytes” by CID, which is a codec reference and a hash of the data. A codec, in this case Capnproto, is responsible for serializing and deserializing those bytes. We know to use the Capnproto codec because of the codec reference in the **link** (CID).

In your codec, you don’t really need to worry about Blocks too much because the codec will over every get one blocks worth of data to serialize or de-serialize at a time. You will need to worry about storing CID’s though, but you shouldn’t only store CID’s that point to Capnproto data, that’s the beauty of CID’s and IPLD, you can have data structures that link to other codecs.

This is all entirely unrelated to how your codec stores Binary data. If you want to take arbitrary bytes as a value within your codec you’ll want to find a way to serialize that. That’s all unrelated to CIDs and Blocks.

**Comment by lthibault** [comment]
@mikeal Thanks for the detailed explanations - I think everything is starting to click into place.

Let me just rephrase the key parts in my own words, to make sure I've understood correctly:

### Link

**At the data-model level:**  a _Link_ is a CID that points to a blob of bytes.  Said blob can be decoded into a _Node_ using the appropriate _Codec_.  The appropriate codec is determined using the CID's [multicodec field](https://github.com/ipld/specs/blob/master/block-layer/CID.md#summary).

**At the library level:** a Link is represented by [ipld.Link](https://godoc.org/github.com/ipld/go-ipld-prime#Link).

### Block

>We literally just brought up in our team call on Monday that the spec doc for Block is “confusing” so my apologies on this one ;)

If I understand correctly, [this block diagram](https://github.com/ipld/specs/blob/master/block-layer/block.md#concept-block) is misleading because it appears that the CID and blob are adjacent in memory.  The intent of the diagram was to illustrate the _concept_ of a block referring to the abstract tuple `(CID, []byte)`.  A better diagram would therefore be:

```
+-----+     +--------------------------------+
| CID | --> | Data                           |
+-----+     +--------------------------------+
```

where `-->` means "points to".  In other words, I shouldn't expect to see an actual block _type_ anywhere in the code; I should instead use the `CID` as a key in a key-value store, such that it points to the blob.

Is this correct?

**Comment by mikeal** [comment]
> At the library level: a Link is represented by ipld.Link.

This is a @warpfork question, but in the codec/format a link is a CID ;)

**Comment by mikeal** [comment]
>  A better diagram would therefore be:

Yes, this is a much better diagram, I’m going to steal this :)

**Comment by lthibault** [comment]
> Yes, this is a much better diagram, I’m going to steal this :)

If ~imitation~ theft is the sincerest form of flattery, consider me flattered 😉

I think it bears emphasizing in the docs that the `-->` is an _arbitrary_ function.  The reason there's no cohesive "block object" in the code is because a block is an abstract relationship formed by (1) a CID (2) some bytes and (3) some mapping function that takes us from 1 to 2.  In IPFS, it's bitswap, but in theory it could be anything.

It's precisely this that was missing from my mental model until now.  I was frantically searching through the repos for some code object called `Block`.  The matter was further confused by the presence of deprecated code using `block.Block`.  FWIW, I wouldn't shy away from repeatedly beating the reader over the head with this point.
